### PR TITLE
bugfix: fix multi-cta top-k implementation when k value is different for different row

### PR DESCRIPTION
**Pull request by yzh119**
## 📌 Description

  This PR fixes a bug in the multi-CTA radix top-k kernel where histogram data corruption occurs when processing batches with mixed k values (some k >= vocab_size, others k < vocab_size) in persistent loop mode.

  Root Cause: When a row has k >= vocab_size, the radix select phase is skipped entirely. However, this leaves stale histogram data in the RadixRowState buffer. When a subsequent row in the same persistent loop iteration has k < vocab_size, it uses atomicAdd on these stale histograms, causing incorrect pivot selection and wrong top-k results.

  Conditions to trigger the bug:
  1. Multi-CTA mode (vocab_size > 57344)
  2. Persistent loop mode (batch_size > num_groups)
  3. Mixed k values across rows (some k >= vocab_size, others k < vocab_size)

  Changes

  - include/flashinfer/topk.cuh:
    - RadixSelectFromSharedMemory: Only clear histogram[0] when iter == 0
    - RadixTopKKernel_Unified: Add histogram clearing in all k >= length paths (Basic, PageTableTransform, RaggedTransform modes)
    - RadixTopKMaskLogitsKernel_MultiCTA: Add histogram clearing in k >= vocab_size path
    - RadixTopKRenormProbKernel_MultiCTA: Add histogram clearing in k >= vocab_size path
  - tests/utils/test_sampling.py:
    - Added regression tests for mixed k values with persistent loop

## 🔍 Related Issues

#2320 

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved inter-worker synchronization and explicit histogram buffer clearing across top-k kernels to prevent cross-iteration/barrier data hazards, improving reliability for large batches and persistent (multi-worker) loops.

* **Tests**
  * Added mixed-k tests for probability renormalization and logit masking across dtypes and large batch/vocabulary sizes to validate persistent/multi-worker behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: include/flashinfer/topk.cuh [hunk 1]
```diff
@@ -499,10 +502,23 @@ __device__ __forceinline__ OrderedType RadixSelectFromSharedMemory(
...
+        for (uint32_t i = tx; i < RADIX; i += BLOCK_THREADS) {
+          state->histogram[0][i] = 0;
+        }
+      }
+      if (tx == 0) {
+        st_release(&state->output_counter, 0);
+      }
     }
+    __syncthreads();  // Ensure histogram clearing completes before any CTA proceeds
   }
```
[the remark this answers is not in the document]
Comment (yzh119) [reply include/flashinfer/topk.cuh]: > If state->histogram[0] can be non-zero at kernel start

They are initialized to 0 at python-side when passing to the kernel at the first time, and we should guarantee they are 0 after kernel finished.

File: tests/utils/test_sampling.py [hunk 2]
```diff
@@ -513,6 +513,83 @@ def test_top_k_renorm_probs(batch_size, vocab_size, k, distribution, dtype):
...
+            topk_vals, _ = torch.topk(row_probs, k, sorted=True)
+            threshold = topk_vals[-1]
+            expected_ge_threshold = (row_probs >= threshold).sum().item()
+
+            # Allow small tolerance for floating point
+            assert count >= k - 1, f"Row {i}: k={k} but only {count} non-zero elements"
+            assert count <= expected_ge_threshold + 1, (
+                f"Row {i}: k={k}, expected at most {expected_ge_threshold} but got {count}"
+            )
+
```
[the remark this answers is not in the document]
Comment (yzh119) [reply tests/utils/test_sampling.py:591]: We have enough GPU memory on CI.

File: tests/utils/test_sampling.py [hunk 3]
```diff
@@ -597,6 +674,68 @@ def test_top_k_mask_logits(
...
+            topk_vals, _ = torch.topk(row_logits, k, sorted=True)
+            threshold = topk_vals[-1]
+            expected_ge_threshold = (row_logits >= threshold).sum().item()
+
+            # Allow small tolerance for floating point
+            assert count >= k - 1, f"Row {i}: k={k} but only {count} finite elements"
+            assert count <= expected_ge_threshold + 1, (
+                f"Row {i}: k={k}, expected at most {expected_ge_threshold} but got {count}"
+            )
+
```
[the remark this answers is not in the document]
Comment (yzh119) [reply tests/utils/test_sampling.py:737]: ditto.

File: include/flashinfer/topk.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (yzh119) [reply include/flashinfer/topk.cuh]: After a second thought, I suppose the change we made here is redundant.
